### PR TITLE
Expand log terms into prime factors.

### LINKED ISSUE #18674: rewrite log(12)/log(2) as 2 + log(3)/log(2)
I am not sure how to rewrite the quotient of logs to remove powers of the denominator's factors. I thought it could be recognized in `logcombine` but that function already does this
```python
>>> logcombine(log(12)/log(2))
log(12**(1/log(2)))
```
So maybe this is something that should be done by `expand_log` with a keyword `factor` is used to factor arguments before running.
```
>>> (log(12)/log(2)).replace(
...    lambda x: x.func is log and x.args[0].is_Integer, 
...    lambda x: log(factorint(x.args[0], visual=True)))
log(2**2*3**1)/log(2**1)
>>> expand_log(_)
(log(3) + 2*log(2))/log(2)
>>> expand(_)
log(3)/log(2) + 2
```
Not sure...seems like there should be an easy way to expand logs of integers.

### PR DESCRIPTION
**Pull request by sbt4104**
#### References to other Issues or PRs
Fixes #18674

#### Brief description of what is fixed or changed
expand log term into its prime factors if factor flag is True
example:
`x = log(12)/log(2)`
`expand(x, factor=True) = log(3)/log(2) + 2
`

### Release Notes

* functions
    - expand log term into its prime factors if the factor flag is True.

### CODE REVIEW

File: sympy/functions/elementary/exponential.py [hunk 1]
```diff
@@ -753,8 +753,8 @@ def taylor_term(n, x, *previous_terms):  # of log(1+x)
                 return powsimp((-n) * p * x / (n + 1), deep=True, combine='exp')
         return (1 - 2*(n % 2)) * x**(n + 1)/(n + 1)
 
-    def _eval_expand_log(self, deep=True, **hints):
-        from sympy import unpolarify, expand_log
+    def _eval_expand_log(self, deep=True, factor=False, **hints):
```
Comment (smichr) [inline sympy/functions/elementary/exponential.py]: I suspect this should be passed via `hints`

Comment (sbt4104) [reply sympy/functions/elementary/exponential.py]: please check the changes.

**Comment by smichr** [comment]
Maybe the simplify_basuc method could look for ratios of logs and just handle factors in common between them. This would be easier than factoring...and factoring won't gain any advantage unless there are logs being subtracted.

**Comment by asmeurer** [comment]
Why is a flag needed? Shouldn't expand_log always do this?

File: sympy/functions/elementary/exponential.py [hunk 2]
```diff
@@ -765,6 +766,11 @@ def _eval_expand_log(self, deep=True, **hints):
             p = perfect_power(int(arg))
             if p is not False:
                 return p[1]*self.func(p[0])
+            # expand as product of its prime factors if factor=True
+            if factor:
+                p = factorint(int(arg))
+                if int(arg) not in p.keys():
+                    return sum(log(arg / val**n) for val, n in p.items())
```
Comment (asmeurer) [inline sympy/functions/elementary/exponential.py]: Is this formula right?

Comment (sbt4104) [reply sympy/functions/elementary/exponential.py]: sorry, It should be 
`sum(n*log(val) for val, n in p.items())`

this should also remove the need for previous condition
`p = perfect_power(int(arg))`
`if p is not False: return p[1]*self.func(p[0])`

**Comment by sbt4104** [comment]
> Why is a flag needed? Shouldn't expand_log always do this?

I think it is cheaper to just put in log together if factors are not needed as it involves factorization, there is also this test in test_exponential

`assert log(2*3**2).expand() != 2*log(3) + log(2)`

**Comment by smichr** [comment]
> Shouldn't expand_log always do this?

I'm not sure that attempting factoring for all arguments is a good idea for default. So for that reason I think it's good to have as a flag. But I also think that we should could add an `expand_basic` for `Mul` that would look for ratios of logs to select the factor to extract so `log(4*bigoddmaybeprime)/log(2)` would become `2 + log(bigoddmaybeprime)/log(2)` without attempting to factor that big number.

**Comment by sbt4104** [comment]
>`log(4*bigoddmaybeprime)/log(2)` would become `2 + log(bigoddmaybeprime)/log(2)` without attempting to factor that big number.

Terms like `log(x)/log(y)` should be returned as `log(x,y)` which already handles the powers of base.

File: sympy/functions/elementary/exponential.py [hunk 3]
```diff
@@ -762,10 +762,15 @@ def _eval_expand_log(self, deep=True, **hints):
             return expand_log(self.func(*self.args), deep=deep, force=force)
         arg = self.args[0]
         if arg.is_Integer:
-            # expand log as product of its prime factors
-            p = factorint(int(arg))
-            if int(arg) not in p.keys() and factor:
-                return sum(n*log(val) for val, n in p.items())
+            # remove perfect powers
```
Comment (smichr) [inline sympy/functions/elementary/exponential.py]:
```
            p = perfect_power(int(arg))
            logarg = None
            coeff = 1
            if p is not False:
                arg, coeff = p
                logarg = self.func(arg)
            # expand as product of its prime factors if factor=True
            if factor:
                p = factorint(int(arg))
                if int(arg) not in p.keys():
                    logarg = sum(n*log(val) for val, n in p.items())
            if logarg is not None:
                return coeff*logarg
```

**Comment by smichr** [comment]
> which already handles the powers of base

For `log(a)*log(b)/log(c)/log(d)` some introspection would have to be done to see which values can be fortuitously combined `log(12)/log(2)*log(18)/log(3) -> 2*log(2)/log(3) + 2*log(3)/log(2) + 5`

File: sympy/functions/elementary/exponential.py (hunk 3 above)
Comment (sbt4104) [reply sympy/functions/elementary/exponential.py]: I will make this change.

**Comment by smichr** [comment]
> some introspection would have to be done

I recall working on the log rules for simplification and part of what drove the decisions then was that it's really only some hollow simplification that can result by doing these sorts of transformations. But...if the user wants to do it, now there is a way to do it. And it could always be followed up with `logcombine`  to change `expanded(log(factorial(5))/log(2), factor=True)` from `log(3)/log(2) + log(5)/log(2) + 3` to `3 + log(15**(1/log(2)))` and another `expand` without factor to give `3 + log(15)/log(2)`.

File: sympy/functions/elementary/exponential.py (hunk 3 above)
Comment (smichr) [reply sympy/functions/elementary/exponential.py]: Having second thoughts...you could just check the for the `coeff = igcd(p.values())` and if it isn't 1 then all values can be divided by `coeff`.

Comment (sbt4104) [reply sympy/functions/elementary/exponential.py]: I didn't understand the use of `igcd(p[0],p[1])`, 
if p = (3,3), then `log(27)/3` wont be solved further

**Comment by sbt4104** [comment]
> another `expand` without factor to give `3 + log(15)/log(2)`.

should I make changes in Mul so that it identifies ratios `log(x)/log(y)` of log and convert them into `log(x,y)`, this way we can get `log(x)/log(y) = log(x,y)` without the need to expand it.

**Comment by smichr** [comment]
> without the need to expand it.

I am not sure. What is the purpose of making such changes? In theory things can be complicated but in real-world applications I am not sure how often these sorts of things will arise and how this rewriting will make things simpler.

**Comment by smichr** [comment]
> should I make changes in Mul so that it identifies ratios

another way to handle this is with `expand_log` when the `factor` flag is False: 

```python
>> expand_log(log(15)/log(3) + log(4))
1 + log(5)/log(3) + log(4)
>> expand_log(_, factor=True)
1 + 2*log(2) + log(5)/log(3)
```

Internally, the following would be used:
```python
...
if factor is False:
    expr = expr.replace(
    lambda x: x.is_Mul and all(any(isinstance(i, log) and i.args[0].is_Rational for i in Mul.make_args(j)) for j in 
    x.as_numer_denom(),
    lambda x: expand_mul(expand_log(x, factor=True)))
...
```

**Comment by sbt4104** [comment]
I managed to accidentally delete the branch I initiated the request from.
should I send another PR for this, please help?

**Comment by smitgajjar** [comment]
Did you delete local branch or master branch?

**Comment by sbt4104** [comment]
I deleted the master branch

**Comment by smitgajjar** [comment]
Do you have that same branch on your system? If yes, then simply push it to the forked repository. 
`git push origin master`
After this, you will have branch having same name and changes.

**Comment by smichr** [comment]
git keeps track of your work; if you do `git reflog` it will print the history of commits. Say you see your last commit with sha `a8d958b`. You just have to do:
```
git checkout a8d958b
git checkout -b repo-name
git push your-name repo-name
```
But I see that at the top of this page "unknown repository" is showing so this might not work.

**Comment by sbt4104** [comment]
I will send a new PR for this.

**Comment by sbt4104** [comment]
As far as I understand the issue, we want to expand the log terms, that we are doing by using factor=True, which expands every log term into its prime factors.
right now log() has the following inconsistency
`log(15,3) = 1 + log(5)/log(3)`
this form takes care of the power of base without using expand.

whereas
`log(15)/log(3) = log(15)/log(3)`

though using `factor` will a good addition to the log module where the user wants to expand expression into its simplest form, 
I think that in many cases further simplifications are stuck due to the non-recognisition of `log(x)/log(y)` as `log(x,y)`.
example
`log(15)/log(3) + log(4)` can be rewritten as 
`log(15,3) + log(4) = 1 + log(4) + log(5)/log(3)`
recognising such cases in mul may reduce the computation cost.

**Comment by sbt4104** [comment]
@smichr, what should be the approach towards this problem. 
should we have expand with `factor`  act independently while mul handling the ratios, as every expression with the possible simplification of ratios will be processed by mul first and further expansion can be done using `expand_log`.
this will also keep the behaviour of `log(x)/log(y)` same as `log(x,y)`.

**Comment by smichr** [comment]
I think an enhancement to `expand_log` that is done even when `factor` is False (as described [here](https://github.com/sympy/sympy/pull/18865#issuecomment-602080143)) is the way to go. I am open to discussion, however.

**Comment by sbt4104** [comment]
I will also go with this, I will make changes and create a PR.

**Comment by smichr** [comment]
moved to #18927

**Pull request closed without merging**